packages/flame_devtools/lib/widgets/component_tree_model.dart

```diff
@@ -0,0 +1,93 @@
+import 'package:animated_tree_view/animated_tree_view.dart';
+import 'package:flame/devtools.dart';
+import 'package:flame_devtools/repository.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+final selectedTreeNodeProvider =
+    StateProvider<TreeNode<ComponentTreeNode>?>((_) => null);
+
+// This is in a separate provider due to the fact that the tree can't keep
+// track of whether it has been changed or not since the nodes don't implement
+// the == operator and hashCode properly.
+final loadedTreeModelProvider = StateProvider<ComponentTreeModel>(
+  (_) => ComponentTreeModel(),
+);
+
+final componentTreeLoaderProvider = FutureProvider<void>((ref) async {
+  final previousTreeModel = ref.watch(loadedTreeModelProvider);
+  final updatedModel =
+      await ComponentTreeModel.refreshComponentTree(previousTreeModel);
+  if (updatedModel != null) {
+    ref.read(loadedTreeModelProvider.notifier).state = updatedModel;
+  }
+});
+
+@immutable
+class ComponentTreeModel {
+  ComponentTreeModel({
+    this.totalId = 0,
+    this.componentCount = 0,
+    TreeNode<ComponentTreeNode>? treeRoot,
+  }) : treeRoot = treeRoot ?? TreeNode<ComponentTreeNode>.root();
+
+  final TreeNode<ComponentTreeNode> treeRoot;
+  final int componentCount;
+  final int totalId;
```

Comment:
basically these are used to detect changes in some way?

Reply:
The `totalId` yes, it's basically a hash that represents the state of the tree.